src/shared_gui_components/BuildingComponentDialogCentralWidget.cpp

```diff
@@ -169,53 +168,87 @@ void BuildingComponentDialogCentralWidget::setTid() {
+    if (filterType == "components") {
+      pageResponses = remoteBCL.searchComponentLibrary(searchString.toStdString(), tid, currentPage);
+    } else if (filterType == "measures") {
+      pageResponses = remoteBCL.searchMeasureLibrary(searchString.toStdString(), tid, currentPage);
+    }
+    responses.insert(responses.end(), pageResponses.begin(), pageResponses.end());
+    totalPages = remoteBCL.numResultPages();
+  } while (++currentPage < totalPages);
+
+  if (!responses.empty()) {
+    std::sort(responses.begin(), responses.end(), [](const BCLSearchResult& a, const BCLSearchResult& b) {
+      return a.name() < b.name();
+    });
+  }
+
+  return responses;
+}
+
 // Note: don't call this directly if the "wait" screen is desired
 void BuildingComponentDialogCentralWidget::setTid(const std::string& filterType, int tid, int pageIdx, const QString& title,
                                                   const QString& searchString) {
 
-  if (m_tid != tid || m_searchString != searchString) {
-    m_collapsibleComponentList->firstPage();
-  }
+  std::string newKey = std::to_string(tid) + filterType + searchString.toStdString();
+  std::string currentKey = std::to_string(m_tid) + m_filterType + m_searchString.toStdString();
 
-  m_filterType = filterType;
+  m_collapsibleComponentList->firstPage();
 
-  m_tid = tid;
+  if (newKey != currentKey) {
+    m_allResponses = fetchAndSortResponses(filterType, tid, searchString);
+  }
 
   m_searchString = searchString;
+  m_filterType = filterType;
+  m_tid = tid;
 
-  //std::vector<Component *> components = m_collapsibleComponentList->components();
+  // Clear existing components
   std::vector<Component*> components = m_componentList->components();  // TODO replace with code above
-
   for (auto& comp : components) {
     delete comp;
   }
 
-  RemoteBCL remoteBCL;
-  remoteBCL.setTimeOutSeconds(m_timeoutSeconds);
-  std::vector<BCLSearchResult> responses;
-  if (filterType == "components") {
-    responses = remoteBCL.searchComponentLibrary(searchString.toStdString(), tid, pageIdx);
-  } else if (filterType == "measures") {
-    responses = remoteBCL.searchMeasureLibrary(searchString.toStdString(), tid, pageIdx);
-  }
-
-  for (const auto& response : responses) {
-    auto* component = new Component(response);
-
-    // TODO replace with a componentList owned by m_collapsibleComponentList
-    m_componentList->addComponent(component);
+  // Paginate responses
+  int itemsPerPage = 10;  // Assuming 10 items per page
+  
+  if (!m_allResponses.empty()) {
+    size_t startIdx = pageIdx * itemsPerPage;
+    size_t endIdx = std::min(startIdx + itemsPerPage, m_allResponses.size());
+    std::vector<BCLSearchResult> paginatedResponses(m_allResponses.begin() + startIdx, m_allResponses.begin() + endIdx);
+  
+    for (const auto& response : paginatedResponses) {
+      auto* component = new Component(response);
+  
+      // TODO replace with a componentList owned by m_collapsibleComponentList
```

Comment:
What is this TODO about?

Reply:
This was in the original code, but it appears as added here

`m_componentList->addComponent(component);`

I thought of removing it but maybe we should just keep it as we are keeping 

`m_componentList->addComponent(component);`